### PR TITLE
BUG: avoid assert statements in business logic (`io.ascii`)

### PR DESCRIPTION
**Pull request by neutrinoceros**
### Description
ref #17472 

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Comment by neutrinoceros** [comment]
Note that there are still 4 failing tests with
```
PYTHONOPTIMIZE=2 pytest astropy/io/ascii
```
but they all have the same, orthogonal failure mode:
`astropy.io.registry.interface.UnifiedReadWrite.help` relies on accessing `__doc__` attributes at runtime
https://github.com/astropy/astropy/blob/7c800f89fe06672a330bba82eb9fd957a7915e1b/astropy/io/registry/interface.py#L56
I don't see an easy way out other than skipping these tests under optimized mode; the alternative being to refactor every method whose `__doc__` attribute can be hit there and move the docstring to *another* attribute (the docstring would still be aliased to it, of course); this would be a substantial effort (probably not worth the expected gain), and wouldn't even work for external subclasses.

**Review by mhvk — COMMENTED** [review]
Good to get rid of the functional `assert`s, but may as well aim to make the code clearer...

File: astropy/io/ascii/ui.py [hunk 1]
```diff
@@ -289,7 +289,8 @@ def is_ducktype(val, cls):
                 # See if ``val`` walks and quacks like a ``cls```.
                 try:
                     new_val = cls(val)
-                    assert new_val == val
+                    if not new_val == val:
```
[the remark this answers is not in the document]
Comment (neutrinoceros) [reply astropy/io/ascii/ui.py:292]: much better, thanks !

**Review by mhvk — APPROVED** [review]
Ah, I guess the `np.dtype` was meant for things like `float`, `np.int64` etc.

Looks good to me now - this obviously does the same as what was there before, so let's get it in.

**Pull request merged**